### ISSUE TITLE
Issues with escaping/handling of JSON log with double quotes, special characters and multilined log

### ISSUE THREAD

**Issue by kiich** [fluent/fluent-bit#1370]
## Bug Report

**Describe the bug**
I have been reading up on pretty much all the fluent-bit JSON parsing issues logged, both Open and Closed and have, as far as I can tell tried all the combination of decoders that is available to no solution that fixes all of my problem so I am raising an issue.

NOTE - this is pretty similar if not the same as issue https://github.com/fluent/fluent-bit/issues/1278 which has been excellently written up and I wanted to add more data/context to it with my findings.
Also issue https://github.com/fluent/fluent-bit/issues/615 was what made me believe the problem I was facing originally is fixed HOWEVER the fix to that issue introduced a brand new problem which I'd like to go into detail here.

The problem in a nut shell is:
If docker/container log line contains the following pattern ALL in 1 go, no decoder combination will be able to handle ALL of them and display the result correctly in Splunk that can handle JSON normally fine.

NOTE all of the log line below are direct from docker log so escaping has been done already by it.

**My log lines**
1. double quotes inside a JSON key-value pair's value field. i.e. 
App logs like this:
```
{"message": "target="ckoGraphiteMetricReporter_report" duration=442 result=SUCCESS"}
```
And the resulting docker log looks like this:
```
{"log":"{\"timestamp\":\"2019-06-03T20:41:27.292+00:00\",\"level\":\"INFO\",\"message\":\"target=\\\"ckoGraphiteMetricReporter_report\\\" duration=442 result=SUCCESS\"}\n", "stream":"stdout","time":"2019-06-03T20:41:27.292925629Z"}
```

2. special character such as double byte (Japanese/Chinese and so on)
again, docker log looks like this:
```
{"log":"{\"timestamp\":\"2019-06-03T20:41:27.292+00:00\",\"level\":\"INFO\",\"message\":
\\\"destinationName\\\":\\\"Markotabödöge, Győr, Hungary\\\",\\\"city\\\":\\\"上海市\\\",\\\"country\\\":\\\"الرياض\\\"\"}\n", "stream":"stdout","time":"2019-06-03T20:41:27.292925629Z"}
```

3. Multilined log line such as java stack trace
again, docker log looks like this:
```
{"log":"{\"timestamp\":\"2019-06-03T20:41:27.292+00:00\",\"level\":\"INFO\",\"message\":\"org.xml.sax.SAXParseException: Premature end of file.\\n\\tat org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(Unknown Source)\\n\\tat org.apache.xerces.util.ErrorHandlerWrapper.fatalError(Unknown Source)\\n\\tat org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)\\n\\tat org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)\\n\\tat org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)\\n\\tat org.apache.xerces.impl.XMLVersionDetector.determineDocVersion(Unknown Source)\"}\n", "stream":"stdout","time":"2019-06-03T20:41:27.292925629Z"}
```

**I will now go into decoders combinations I have tried**

1. Using Decoders config combination below:
```
    [PARSER]
        Name        docker
        Format      json
        Time_Key    time
        Time_Format %Y-%m-%dT%H:%M:%S.%L
        Time_Keep   Off
        # Command      |  Decoder     | Field | Optional Action
        # =============|======================|=================
        Decode_Field_As   escaped       log    do_next
        Decode_Field      json  log
```
Fixes log lines 2 & 3 - so special characters gets displayed correctly and multilined log line are displayed as multiple lines in Splunk.

**HOWEVER**, it does not recognise log line 1 as valid JSON because I assume it sees the double quote inside double quote as invalid JSON - the result of this is `message` does not get recognised as query-able field in Splunk.

2. Using Decoders config combination below:
```
    [PARSER]
        Name        docker
        Format      json
        Time_Key    time
        Time_Format %Y-%m-%dT%H:%M:%S.%L
        Time_Keep   Off
        # Command      |  Decoder     | Field | Optional Action
        # =============|======================|=================
        Decode_Field_As   escaped_utf8       log    do_next
        Decode_Field      json  log
```

Fixes log line 1 - double quotes inside double quotes gets escaped correctly it seems and Splunk sees it as valid JSON and the field `message` is now query-able.

**HOWEVER** this decoder config breaks log lines 2 and 3 - so the special character is no longer displayed/recognised and the multilined log line also breaks because it displays `\\n` as a literal string in Splunk.

3. Using Decoders config combination below:
```
    [PARSER]
        Name        docker
        Format      json
        Time_Key    time
        Time_Format %Y-%m-%dT%H:%M:%S.%L
        Time_Keep   Off
        # Command      |  Decoder     | Field | Optional Action
        # =============|======================|=================
        Decode_Field_As   json       log
```
Fixes 1 and 2 - so double quotes inside double quotes gets escaped ok and the filed `message` is query-able as well as special character displayed correctly.

**HOWEVER** multilined log line is not displayed as multilines.

I noticed in the issue 1278, using decoder json actually does:
> One important aspect of this decoder (and not documented anywhere as far as I can tell) is that it first does a escaped decode on the string before attempting to parse it.

which might explain why the above is happening?

I have also tried different combination as per issue 1278 describes, such as running `escaped_utf8` and then `escaped` but nothing seems to be fix ALL of the log lines i have all in 1 go.

**Comment by kiich** [comment]
Just to add more on the last decoder config I used.

I've also tried to follow the 
```
Decode_Field_As   json       log
```
with further decoder i.e.
```
        Decode_Field_As   json          log do_next
        Decode_Field_As      escaped_utf8       log
```
as well as
```
        Decode_Field_As   json          log do_next
        Decode_Field_As      escaped      log
```
but they do not seem to make any difference what so ever on the result in Splunk display. I guess this is because of:
> on Decode_Field_As, if succeeded, another decoder of the same type in the same field can be applied only if the data continue being a unstructed message (raw text).

so after the first `json` decoder, it is no longer a raw text hence subsequent decoder does not do anything maybe?

**Comment by kiich** [comment]
Comment to say that fix mentioned in #1278 has fixed all of my log lines issues mentioned in this ticket if anybody comes here who are running into similar issues as me.
tl;dr; - fix is coming as official release hopefully soon!

**Comment by edsiper** [comment]
closing this as fixed, for now, the encoding conversation can continue on #1278.

**Comment by kiich** [comment]
Yep, sorry @edsiper I should have closed this myself. Cheers.
